statsd-client/statsd_client_core.ml

```diff
@@ -51,83 +65,77 @@ module Make (U : IO) :
-              socket_ref := Some s;
-              s
+          | Some s -> s
+          | None   ->
+            !log_debug "Creating statsd socket";
+            let s = U.socket Unix.PF_INET Unix.SOCK_DGRAM proto in
+            socket_ref := Some s;
+            s
         in
         let portaddr = Unix.ADDR_INET (Unix.inet_addr_of_string ipaddr, port) in
-        (* Map sample_rate to the string statsd expects *)
-        let sample_rate =
-          if sample_rate >= 1.0 then
-            ""  (* not sampling *)
-          else
-            Printf.sprintf "|@%f" sample_rate
-        in
         U.list_iter
-          (fun (stat, value) ->
-            let msg = Printf.sprintf "%s:%s%s" stat value sample_rate in
-            U.sendto socket msg 0 (String.length msg) [] portaddr
-            >>= (fun retval ->
-              if retval < 0 then
-                begin
-                  (* Ran into an error, clear the reference
-                     and close the socket *)
-                  socket_ref := None;
-                  U.catch
-                    (fun () ->
-                      !log_debug
-                        (Printf.sprintf "Closing statsd socket: %d" retval);
-                      U.close socket >>= U.return)
-                    (fun _e -> U.return ());
-                end
-              else U.return ()
-            )
+          (fun msg ->
+             U.sendto socket msg 0 (String.length msg) [] portaddr
+             >>= (fun retval ->
+                 if retval < 0 then
+                   begin
+                     (* Ran into an error, clear the reference
+                        and close the socket *)
+                     socket_ref := None;
+                     U.catch
+                       (fun () ->
+                          !log_debug
+                            (Printf.sprintf "Closing statsd socket: %d" retval);
+                          U.close socket >>= U.return)
+                       (fun _e -> U.return ());
+                   end
+                 else U.return ()
+               )
           )
           data
-      ) (fun _e -> U.return ()
-      )
-    else U.return ()
+      ) (fun _e -> U.return ())
 
-  let send ?(sample_rate = 1.0) data =
+  let send ~fmt ~data =
     match U.ipaddr (), U.port () with
-      | None, _ | _, None ->
-        !log_error
-          "Statsd_client.send: \
-           uninitialized Statsd_client.host or Statsd_client.port";
-        U.return ()
-      | Some ipaddr, Some port ->
-        send_with_ipaddr_and_port ipaddr port sample_rate data
+    | None, _ | _, None ->
+      !log_error
+        "Statsd_client.send: \
+         uninitialized Statsd_client.host or Statsd_client.port";
+      U.return ()
+    | Some ipaddr, Some port ->
+      send_with_ipaddr_and_port ipaddr port (fmt data)
+
+  let fmt_data ?(sample_rate = 1.0) data =
+    if sample_rate >= 1.0 || sample_rate >= Random.float 1.0
```

Comment:
I remember you told me in person, but what was the reasoning for moving this logic into the `fmt_data` fn from the `send` fn?

Reply:
* the send function isn't dependent upon the sample rate like it was prior to this revision
* check that our ip and port are set prior to formatting our data
I've refactored the code since we chatted and as a result, there really isn't a compelling reason anymore. This might be something that can be refactored out as its no longer that useful and at best is a small optimization.
Here is an example of this being used in dogstatsd: https://github.com/arenadotio/ocaml-statsd-client/pull/4/commits/5ab42c1806ecc69c4b8996239d3d0acdd23b1546#diff-ee22f142e5f6184713e7201d0eccc1a0R147

I don't feel strongly about it but do you think I should refactor the code to be such that the signature of `send` is:
```
val send: string list -> unit _r
```
Where `string list` is the formatted data?

Reply:
Haha, I was wondering if we needed to pass `~fmt` to this fn but didn't know how we were planning to use it in the future.